src/lib.rs

```diff
@@ -1,808 +1,268 @@
+    /// The default logger discards all records passed to it.
+    pub fn logger(&self) -> slog::Logger {
+        self.log.clone()
     }
+}
 
-    /// Spin up a tsunami batching the defined machine sets in this builder with a custom
-    /// credentials provider.
+impl<M: MachineSetup + Clone> TsunamiBuilder<M> {
+    /// Add multiple machine descriptors to the Tsunami.
     ///
-    /// When all instances are up and running, the given closure will be called with a handle to
-    /// all spawned hosts. When the closure exits, the instances are all terminated automatically.
+    /// This is a convenience wrapper around [`add`](TsunamiBuilder::add).
     ///
+    /// The `nickname_prefix` is used to name the machines, indexed from 0 to `n`:
     /// ```rust,no_run
-    /// # extern crate rusoto_core;
-    /// # extern crate rusoto_sts;
-    /// # extern crate tsunami;
-    /// # fn main() {
-    /// # use tsunami::{TsunamiBuilder, Machine};
-    /// # use std::collections::HashMap;
-    /// // https://github.com/rusoto/rusoto/blob/master/AWS-CREDENTIALS.md
-    /// let sts = rusoto_sts::StsClient::new(rusoto_core::Region::UsEast1);
-    /// let provider = rusoto_sts::StsAssumeRoleSessionCredentialsProvider::new(
-    ///     sts,
-    ///     "arn:aws:sts::1122334455:role/myrole".to_owned(),
-    ///     "session-name".to_owned(),
-    ///     None,
-    ///     None,
-    ///     None,
-    ///     None,
-    /// );
+    /// fn main() -> Result<(), failure::Error> {
+    ///     use tsunami::providers::Launcher;
+    ///
+    ///     let m = tsunami::providers::aws::Setup::default();
+    ///     let mut aws: tsunami::providers::aws::Launcher<_> = Default::default();
+    ///     let mut b = tsunami::TsunamiBuilder::default();
+    ///     b.add_multiple(3, "my_tsunami", m)?.spawn(&mut aws)?;
     ///
-    /// let mut b = TsunamiBuilder::default();
-    /// // ...
-    /// b.run_as(provider, |vms: HashMap<String, Vec<Machine>>| {
-    ///     println!("==> {}", vms["server"][0].private_ip);
-    ///     for c in &vms["client"] {
-    ///         println!(" -> {}", c.private_ip);
-    ///     }
+    ///     let vms = aws.connect_all()?;
+    ///     let my_first_vm = vms.get("my_tsunami-0").unwrap();
+    ///     let my_last_vm = vms.get("my_tsunami-2").unwrap();
     ///     Ok(())
-    /// }).unwrap();
-    /// # }
-    pub fn run_as<P, F, R>(self, provider: P, f: F) -> Result<R, Error>
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        <P as ProvideAwsCredentials>::Future: Send,
-        F: FnOnce(HashMap<String, Vec<Machine>>) -> Result<R, Error>,
-    {
-        use rusoto_ec2::Ec2;
-
-        let log = &self.log;
-        let rng = rand::thread_rng();
-
-        debug!(log, "connecting to ec2");
+    /// }
+    /// ```
+    pub fn add_multiple(
+        &mut self,
+        n: usize,
+        nickname_prefix: &str,
+        m: M,
+    ) -> Result<&mut Self, Error> {
+        (0..n)
```

Comment:
I think you can just use `std::iter::repeat(m).take(n)` here. Might simplify the code a little.